[1.20.1] Parallelize background file scan

embeddedt opened the pull request:
Backport of https://github.com/neoforged/FancyModLoader/pull/32 to 1.20.1.

Technici4n commented:
What about the scan result future change? Also, is this a breaking change for language loaders?

embeddedt commented:
> What about the scan result future change?

I missed that, added the patch.

> Also, is this a breaking change for language loaders?

I'm very skeptical it would matter, but I added a `synchronized` block in the latest commit so any individual loader's file visitor will never be running concurrently on multiple files. This should be good enough.

Technici4n commented on loader/src/main/java/net/minecraftforge/fml/loading/moddiscovery/Scanner.java:
```diff
@@ -34,7 +34,11 @@ public ModFileScanData scan() {
-                loader.getFileVisitor().accept(result);
+                // Since we now scan files concurrently, but language loaders may not have been written with that in mind,
+                // synchronize on the language loader so it only scans one file at a time
+                synchronized (loader) {
```
Doesn't this negate most of the benefits?

embeddedt replied:
No, the hot code is `Scanner#fileVisitor`, which still runs concurrently.
